// ctx-zip/dist/mcp-sandbox/vercel-sandbox-provider.js
// Vercel Sandbox provider implementation
import { Sandbox } from "@vercel/sandbox";
/**
 * Vercel Sandbox provider implementation
 */
export class VercelSandboxProvider {
    sandbox;
    workspacePath = "/vercel/sandbox";
    constructor(sandbox) {
        this.sandbox = sandbox;
    }
    /**
     * Create a new Vercel sandbox instance
     */
    static async create(options = {}) {
        console.log(`✓ Creating Vercel Sandbox (runtime: ${options.runtime || "node22"})`);
        const sandbox = await Sandbox.create({
            timeout: options.timeout || 1800000,
            runtime: options.runtime || "node22",
            resources: {
                vcpus: options.vcpus || 4,
            },
        });
        console.log("✓ Sandbox created");
        return new VercelSandboxProvider(sandbox);
    }
    async writeFiles(files) {
        await this.sandbox.writeFiles(files);
    }
    async runCommand(command) {
        return await this.sandbox.runCommand({
            cmd: command.cmd,
            args: command.args,
        });
    }
    async stop() {
        await this.sandbox.stop();
    }
    getId() {
        // Vercel sandbox doesn't expose an ID, so we create a simple identifier
        return `vercel-sandbox-${Date.now()}`;
    }
    getWorkspacePath() {
        return this.workspacePath;
    }
    /**
     * Get the underlying Vercel sandbox instance (for advanced use cases)
     */
    getVercelSandbox() {
        return this.sandbox;
    }
}
