// ctx-zip/dist/storage/file.js
import fs from "node:fs";
import { readFile as fsReadFile, writeFile as fsWriteFile, } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
export class FileStorageAdapter {
    baseDir;
    prefix;
    constructor(options) {
        this.baseDir = options.baseDir;
        this.prefix = options.prefix ?? "";
    }
    resolveKey(name) {
        const safe = name.replace(/\\/g, "/").replace(/\.+\//g, "");
        return this.prefix ? `${this.prefix.replace(/\/$/, "")}/${safe}` : safe;
    }
    async write(params) {
        const fullPath = path.resolve(this.baseDir, params.key);
        await fs.promises.mkdir(path.dirname(fullPath), { recursive: true });
        const body = typeof params.body === "string" ? params.body : Buffer.from(params.body);
        await fsWriteFile(fullPath, body, "utf8");
        const url = new URL(`file://${fullPath}`);
        return { key: params.key, url: url.toString() };
    }
    async readText(params) {
        const fullPath = path.resolve(this.baseDir, params.key);
        return await fsReadFile(fullPath, "utf8");
    }
    async openReadStream(params) {
        const fullPath = path.resolve(this.baseDir, params.key);
        return fs.createReadStream(fullPath);
    }
    toString() {
        return `file://${this.baseDir}${this.prefix ? "/" + this.prefix : ""}`;
    }
}
export function fileUriToOptions(uri) {
    // Expect file:///abs/path or file:/abs/path
    const url = new URL(uri);
    if (url.protocol !== "file:") {
        throw new Error(`Invalid file URI: ${uri}`);
    }
    const baseDir = fileURLToPath(url);
    return { baseDir };
}
