// ctx-zip/dist/storage/vercelBlob.js
import { head, put } from "@vercel/blob";
import { Readable } from "node:stream";
export class VercelBlobStorageAdapter {
    prefix;
    access;
    constructor(options = {}) {
        this.prefix = options.prefix ?? "";
        this.access = options.access ?? "public";
    }
    resolveKey(name) {
        const safe = name.replace(/\\/g, "/").replace(/\.+\//g, "");
        return this.prefix ? `${this.prefix.replace(/\/$/, "")}/${safe}` : safe;
    }
    async write(params) {
        const body = typeof params.body === "string" ? params.body : Buffer.from(params.body);
        const result = await put(params.key, body, {
            access: this.access,
            contentType: params.contentType,
            // Ensure the stored object path matches the requested key
            addRandomSuffix: false,
        });
        // Use actual stored pathname as the effective key
        const effectiveKey = typeof result?.pathname === "string" ? result.pathname : params.key;
        return { key: effectiveKey, url: result.url };
    }
    async readText(params) {
        const resolvedKey = this.resolveKey(params.key);
        const meta = await head(resolvedKey);
        const url = meta.downloadUrl || meta.url;
        const res = await fetch(url);
        if (!res.ok)
            throw new Error(`Failed to read blob: ${resolvedKey} (${res.status})`);
        return await res.text();
    }
    async openReadStream(params) {
        const resolvedKey = this.resolveKey(params.key);
        const meta = await head(resolvedKey);
        const url = meta.downloadUrl || meta.url;
        const res = await fetch(url);
        if (!res.ok || !res.body) {
            throw new Error(`Failed to open blob stream: ${resolvedKey} (${res.status})`);
        }
        // Convert Web ReadableStream to NodeJS.ReadableStream
        return Readable.fromWeb(res.body);
    }
    toString() {
        const prefix = this.prefix ?? "";
        if (!prefix)
            return "blob:";
        return `blob://${prefix}`.replace(/\/$/, "");
    }
}
export function vercelBlobUriToOptions(uri) {
    // Support blob root and path forms: blob:, blob:/, blob://, blob:///, blob:/prefix, blob:///prefix
    if (uri === "blob:" ||
        uri === "blob:/" ||
        uri === "blob://" ||
        uri === "blob:///") {
        return {};
    }
    const url = new URL(uri);
    if (url.protocol !== "blob:") {
        throw new Error(`Invalid blob URI: ${uri}`);
    }
    const prefix = url.pathname.replace(/^\//, "");
    return { prefix };
}
